[sparx5] Use correct index into DEV2G5_MAC_MAXLEN_CFG

ievenbach opened the pull request:
DEV2G5_* registers are indexed by port number, not by index within port family (1G vs. 5G vs. 10G vs 25G ports)

This change fixes setting MTU on ports in 1G mode.

ievenbach committed:
[sparx5] Use correct index into DEV2G5_MAC_MAXLEN_CFG

DEV2G5_* registers are indexed by port number, not by
index within port family (1G vs. 5G vs. 10G vs 25G ports)

This change fixes setting MTU on ports in 1G mode.

ievenbach commented:
@HoratiuVultur simple fix to MTU settings

HoratiuVultur commented:
Hi @ievenbach,

I am not sure what is the problem. Because the idx will be port->portno if the device is DEV2G5.
So, the idx gets set here:
https://github.com/microchip-ung/linux/blob/bsp-6.12-2025/drivers/net/ethernet/microchip/sparx5/sparx5_mtu.c#L19
And the return value is calculated here:
https://github.com/microchip-ung/linux/blob/bsp-6.12-2025/drivers/net/ethernet/microchip/sparx5/sparx5_port.c#L1511
And this returns the portno.

Do I miss something?

Thanks,
/Horatiu

ievenbach commented:
> Hi @ievenbach,
> 
> I am not sure what is the problem. Because the idx will be port->portno if the device is DEV2G5.
> So, the idx gets set here:
> https://github.com/microchip-ung/linux/blob/bsp-6.12-2025/drivers/net/ethernet/microchip/sparx5/sparx5_mtu.c#L19
> And the return value is calculated here:
> https://github.com/microchip-ung/linux/blob/bsp-6.12-2025/drivers/net/ethernet/microchip/sparx5/sparx5_port.c#L1511
> And this returns the portno.
> 
> Do I miss something?
> 
> Thanks,
> /Horatiu

Th idx does not get set to portno. In all other places that use sgadow mac, portno is ued directly.

If i understand it correctly, idx is set on port type, not current port mode.

ievenbach commented:
Also, there are lot of places, where that treatment of idx is correct. That is why i changed the call site, and not the function getting the index.

HoratiuVultur commented:
@ievenbach , I was looking at the function sparx5_mtu_get, there the idx is set to portno when the port is_2g5.
But in the function sparx5_mtu_set, you are correct, for a port that is a 10G device then it is using the idx to access the DEV2G5 registers which is wrong.

HoratiuVultur commented:
Hi @ievenbach,

Do you agree that the change needs to be done only in the sparx5_mtu_set function and it is not needed in the function sparx5_mtu_get?

Thanks,
/Horatiu

ievenbach commented:
> Hi @ievenbach,
> 
> Do you agree that the change needs to be done only in the sparx5_mtu_set function and it is not needed in the function sparx5_mtu_get?
> 
> Thanks,
> /Horatiu

I think it's still better for consistency. Also, now that I think about it, since we always keep MTU of 2g5 Mac in sync, get can be simplified to only read that register

ievenbach commented:
> > Hi @ievenbach,
> > Do you agree that the change needs to be done only in the sparx5_mtu_set function and it is not needed in the function sparx5_mtu_get?
> > Thanks,
> > /Horatiu
> 
> I think it's still better for consistency. Also, now that I think about it, since we always keep MTU of 2g5 Mac in sync, get can be simplified to only read that register

I took a closer look. I see handling for rgmii ports in _get, but not set. What is the correct action here?
Irrelevant for sparx5, since it has no RGMII, but relevant for LAN969x, if I understand correctly

HoratiuVultur commented:
Regarding the rgmii ports, I think it is a problem, I will give it a try thanks for spotting it.

ievenbach committed:
[sparx5] Fix RGMII MTU set and cleanup MTU get

- Add RGMII port handling to sparx5_mtu_set()
- Remove extra code from sparx5_mtu_get()

ievenbach commented:
@HoratiuVultur added something (untested). Since I do not know if RGMII port have DEV2G5 MAC, I didn't remove those from _get()

HoratiuVultur commented:
Hi @ievenbach,

Thanks for the latest changes but it doesn't seem to compile anymore.

Thanks,
/Horatiu

ievenbach commented:
> Hi @ievenbach,
> 
> Thanks for the latest changes but it doesn't seem to compile anymore.
> 
> Thanks, /Horatiu

Oops. Fixed. Still untested on real HW - that has to be up to you, since I don't have anything with RGMII ports.

ievenbach commented:
> > Hi @ievenbach,
> > Thanks for the latest changes but it doesn't seem to compile anymore.
> > Thanks, /Horatiu
> 
> Oops. Fixed. Still untested on real HW - that has to be up to you, since I don't have anything with RGMII ports.

Tested on Sparx5i

HoratiuVultur commented:
Thanks for update. Yes, I will try it on lan969x and I will let you know how it works.

HoratiuVultur commented:
Hi @ievenbach,

I have give it a try on lan969x on the RGMII ports and unfortunately it didn't work. The problem is that the RGMII ports don't have a DEV2G shadow device and they need to be accessed using idx and not port->portno.
So, what I have done, I have added a check before setting the DEV2G5 and set it for all ports that are not RGMII ports.
With this change the RGMII ports on lan969x seem to be working.
I will take all these changes and put them in our internal tree.

Thanks for the patch and adding support for RGMII.

Thanks,
/Horatiu
